[IMAGING-339] Basic WebP Support

Glavo opened the pull request:
This PR is intended to provide basic support for the WebP format.

Task List:

- [x] Add WebP support for `Imaging.guessFormat`;
- [x] Read all of webp chunks;
- [x] Implementing `ImageParser::getImageSize` for webp files;
- [x] Implementing `ImageParser::getMetadata` for webp files;
- [x] Implementing `ImageParser::getImageInfo` for webp files.

Non-Goals:

* Read image from webp file;
* Write to webp files.

Although I hope to implement complete WebP support, it is very difficult for me at present.

Therefore, the main goal of this PR is to read metadata from webp files. 

A more complete implementation may require me to have a deeper understanding of vp8 format, or let others complete it.

kinow reviewed (commented):
:tada: looks promising @Glavo ! And pure Java, I thought there would be some dependency here, but much better if we can implement it in Imaging without bringing other jars. Feel free to ping me once it's ready and I can take a look and try it out. Thanks!

Glavo commented:
It seems that there are some strict checkstyle rules here. I try to solve them.

Glavo commented:
@kinow By adding exclusion rules, I no longer get warnings when running `mvn -V --no-transfer-progress` locally. Can you trigger the CI again?

Basic support for WebP has been completed. Now I can successfully read the basic information of the lossy/lossless compressed webp file.

I also implemented dumpFile for WebP files. Now it can print the contents of webp chunks in a similar way to the `webpinfo` tool.

It is difficult for me to implement the function of reading/writing image content. I try to read the source code of libwebp, but it is not easy for me. So I can't make more contributions to WebP support in the near future.

kinow commented:
>@kinow By adding exclusion rules, I no longer get warnings when running mvn -V --no-transfer-progress locally. Can you trigger the CI again?

Done!

>Basic support for WebP has been completed. Now I can successfully read the basic information of the lossy/lossless compressed webp file.
>
>I also implemented dumpFile for WebP files. Now it can print the contents of webp chunks in a similar way to the webpinfo tool.

Thakn you!

>It is difficult for me to implement the function of reading/writing image content. I try to read the source code of libwebp, but it is not easy for me. So I can't make more contributions to WebP support in the near future.

I never read the format docs, so no idea how hard it would be. But one option would be to include it in our upcoming 1.0beta release (one beta before the final release) and document what it does and what it doesn't. It's better to be able to at least read metadata/information, then not being able to read WebP at all.

Thank you @Glavo !

Glavo commented:
@kinow Will the next version be 1.0 beta? Do you have any ideas about when to make AWT as optional dependency? I think it is better to solve this problem in the alpha phase, because this change may break the user's code.

kinow commented:
> @kinow Will the next version be 1.0 beta?

Yes. We could release an 1.0-alpha3 if the API is not stable enough. I was actually thinking in releasing 1.0 final, but there was some internal discussion about it, and it was decided to release at least one beta version after these alpha releases, to a) make sure we have a good public API (i.e. err on having less than what's needed, not more, so that we can maintain it for a while) and b) it has enough features for users.

Do you have any suggestion on alpha/beta/final?

> Do you have any ideas about when to make AWT as optional dependency? I think it is better to solve this problem in the alpha phase, because this change may break the user's code.

I think it is more a matter of deciding the how first. If it is not too complicated, we can include it in the next release.

I can't recall what classes we used from AWT (Dimension maybe?).

We had a user that contributed with a lot of suggestions and by testing pull requests, and his application was being deployed on Android. I think he was the first one to request it, from what I recall, as for the time being he had to do some patching on his code in order to use Imaging.

If you have any suggestions, feel free to let us know, please :slightly_smiling_face: 

BTW, it would be great if you could create a JIRA issue for this issue too, as we need one to track changes in our `changes.xml` (that's updated by committers, no need to worry about that).

Thanks heaps!

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
```diff
@@ -0,0 +1,351 @@
+                hasAlpha = ((WebPChunkVP8X) chunk).isHasAlpha();
+
+                if (vp8x.isAnimation()) {
+                    formatDetails = "WebP/Animation";
```
Worth moving these strings to constants?

Glavo replied:
> Worth moving these strings to constants?

I checked the implementation of other image formats. Format details seems to be an implementation specific internal representation, so I don't think it needs to be placed elsewhere.

kinow replied:
I think these could be constants in the class, like the extensions. Will move them as it's harmless :man_shrugging:

kinow replied:
Oh, we could even use an enum for the formats... but now I see they are being used once, and the performance benefit would be minimal. We can worry about that later. Leaving as-is, resolving this comment.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
```diff
@@ -0,0 +1,351 @@
+                WebPChunkVP8X vp8x = (WebPChunkVP8X) chunk;
+                return new Dimension(vp8x.getCanvasWidth(), vp8x.getCanvasHeight());
+            } else {
+                throw new ImageReadException("Unknown webp chunk type: " + chunk);
```
webp & WebP, to be consistent in the error messages.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
```diff
@@ -0,0 +1,351 @@
+
+                offset = reader.getOffset();
+                chunk = reader.readChunk();
+            } while (chunk != null);
```
Adding a note to self: check for possible loops & add/multiply that could be made more defensive/safe to avoid CVE's (we had a few found by users & by the oss-fuzzer)

kinow replied:
Adding a `TODO` comment. Thought a little about it, and I think we are safe from DOS attacks here.

kinow reviewed (commented):
Hi @Glavo , I did a quick review, and most files look great (I even marked some as "Reviewed" on GitHub).

You have already seen that we have some strict checkstyle/spotbugs rules. These are more or less similar across all ASF Commons projects, as well as the idea to include only what's necessary in the public API, add javadocs to all public code, and to internal code if possible too, etc.

I left some comments about that and about the text in the exceptions, and the exceptions types. But overall it looks great!

Thanks a lot!
-Bruno

Glavo commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
```diff
@@ -0,0 +1,351 @@
+        }
+
+        if (is.read(buffer) < 4 || !WebPConstants.WEBP_SIGNATURE.equals(buffer)) {
+            throw new IOException("Not a Valid WebP File");
```
@kinow I don't understand the real difference between `IOException` and `ImageReadException`. 

I use `IOException` here to imitate the behavior of `BinaryFunctions.read4Bytes`, because it will also throw `IOException` instead of `ImageReadException`.

The subtle difference between them confused me. Should we let `ImagingException` extends `IOException`?

Glavo replied:
@kinow Can you take a look at this question?

Glavo replied:
@kinow Hey?

kinow replied:
(will look at this one later, don't worry :+1: )

kinow replied:
All good over here. The Imaging classes called by users throw `ImagingException`. The `BinaryFunctions` and other classes may raise `IOException`, but that would be either wrapped in an ImagingException or raised directly to the user. At least that was the original design of the component code, if I recall correctly, so we just follow it (more or less). Marking as resolved :+1:

Glavo commented:
> > @kinow Will the next version be 1.0 beta?
> 
> Yes. We could release an 1.0-alpha3 if the API is not stable enough. I was actually thinking in releasing 1.0 final, but there was some internal discussion about it, and it was decided to release at least one beta version after these alpha releases, to a) make sure we have a good public API (i.e. err on having less than what's needed, not more, so that we can maintain it for a while) and b) it has enough features for users.
> 
> Do you have any suggestion on alpha/beta/final?
> 
> > Do you have any ideas about when to make AWT as optional dependency? I think it is better to solve this problem in the alpha phase, because this change may break the user's code.
> 
> I think it is more a matter of deciding the how first. If it is not too complicated, we can include it in the next release.
> 
> I can't recall what classes we used from AWT (Dimension maybe?).
> 
> We had a user that contributed with a lot of suggestions and by testing pull requests, and his application was being deployed on Android. I think he was the first one to request it, from what I recall, as for the time being he had to do some patching on his code in order to use Imaging.
> 
> If you have any suggestions, feel free to let us know, please 🙂
> 
> BTW, it would be great if you could create a JIRA issue for this issue too, as we need one to track changes in our `changes.xml` (that's updated by committers, no need to worry about that).
> 
> Thanks heaps!
> 
> EDIT: BufferedImage too from awt, d'oh

In my understanding, the alpha phase allows large API changes, but after entering the beta phase, the API tends to be stable, and there can be some minor damage, but usually it should not change as much as the alpha phase.

> I can't recall what classes we used from AWT (Dimension maybe?).

Yes, if we provide an alternative to `Dimension`, we can use commons-imaging in more environments without AWT.

However, even if this problem is solved, I think the commons-imaging API still relies too much on AWT.

I don't think so many `BufferedImage` based APIs should be provided, nor should the implementation of each format be based on AWT/Swing. 

Maybe we can provide a general abstraction to hide these details. The implementation of the format should only use a common interface, and behind the interface may be AWT/Swing, JavaFX, Android, or even other file formats writter.

This is my rough idea. Do you have any opinion on it?

kinow commented:
>In my understanding, the alpha phase allows large API changes, but after entering the beta phase, the API tends to be stable, and there can be some minor damage, but usually it should not change as much as the alpha phase.

Correct. That's why it's important to decide on whether it will be an alpha3 or beta1. I didn't have plans to add anything to the public API before 1.0, but I think WebP support could go in any non-final release as long as we take care with what is added to the public API, and add tests/docs.

>However, even if this problem is solved, I think the commons-imaging API still relies too much on AWT.

Agreed. 

>I don't think so many BufferedImage based APIs should be provided, nor should the implementation of each format be based on AWT/Swing.
>
>Maybe we can provide a general abstraction to hide these details. The implementation of the format should only use a common interface, and behind the interface may be AWT/Swing, JavaFX, Android, or even other file formats writter.
>
>This is my rough idea. Do you have any opinion on it?

Hmmm, this sounds interesting!

I just remembered that when I went looking for ways to support Android, I found someone that had a project with the `library-something.jar`, with the interfaces, and `library-something-awt.jar` and `library-something-android.jar`. But I found that solution harder to implement & maintain.

But let's give it another try before the 1.0. I don't mind postponing this release in order to have more alphas/betas. This is the issue for Android, we can rename/use it for AWT: https://issues.apache.org/jira/browse/IMAGING-208 (if you do not have an account, let me know and I'll prepare the sign up process for you, ASF blocked new sign ups due to Spam).

Thanks!!!

-Bruno

Glavo commented:
@kinow Can you open this issue for me? 

I'm sorry to reply you so late, because I'm not feeling well recently (probably infected with COVID-19), and I need a rest for a while.

Besides, I'm not good at English. You can better describe this problem than let me use Google Translate myself.

kinow commented:
> @kinow Can you open this issue for me?
> 
> I'm sorry to reply you so late, because I'm not feeling well recently (probably infected with COVID-19), and I need a rest for a while.
> 
> Besides, I'm not good at English. You can better describe this problem than let me use Google Translate myself.

Hi @Glavo , your English is excellent for me (I am not a native speaker either) , but don't worry I left this notification unread on my GitHub list of notifications, and I will update it later with a JIRA issue.

Get some rest, and I hope you feel better soon.

Thanks
Bruno

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunkXYZW.java:
```diff
@@ -0,0 +1,25 @@
+import org.apache.commons.imaging.ImageReadException;
+
+public final class WebPChunkXYZW extends WebPChunk {
+    public WebPChunkXYZW(int type, int size, byte[] bytes) throws ImageReadException {
```
We need javadocs for public methods at least. If there are internal/private methods, that are not easy to understand (e.g. complex logic, lots of byte handling) it's also recommended to have some javadocs & code comments.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunkVP8.java:
```diff
@@ -0,0 +1,98 @@
+
+        int b0 = bytes[0] & 0xFF;
+        // int b1 = bytes[1] & 0xFF;
+        // int b2 = bytes[2] & 0xFF;
```
I like that these two are commented out, meaning that it was not a mistake to skip these. But we still need a comment, so other devs are aware of why this was done, please.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunk.java:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    public int getChunkSize() {
+        return size + 8 + ((size % 2) == 0 ? 0 : 1);
```
This is a good example of something that's not too complicated, but would be nicer to have a short explanation why we are doing it. If it's in the spec, you can just say so :+1:

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunk.java:
```diff
@@ -0,0 +1,80 @@
+        super(ByteOrder.LITTLE_ENDIAN);
+
+        if (size != bytes.length) {
+            throw new AssertionError("Chunk size must match bytes length");
```
Instead of `AssertionError` I think it would be best to throw the `ImageReadException`

Glavo replied:
@kinow I think `AssertionError` should be used here.

I think `ImageReadException` should be thrown only when there is no error in commons-imaging itself, e.g. we encountered illegal input at runtime.

However, here, illegal input data will not cause this problem. An exception will be thrown only if commons-imaging itself has logical errors, so I use `AssertionError` here.

garydgregory replied:
You must be thinking of IllegalStateException, we should not throw Errors.

kinow replied:
I think it'd be important to be consistent, so users are not surprised when using the API. Have a look at the other parsers, @Glavo , and see what exceptions are being used. If you believe we must change it to another exception type, then we can create a follow up issue to discuss and plan how to do that in all other parsers :+1:

Glavo replied:
> You must be thinking of IllegalStateException, we should not throw Errors.

In my understanding, here I am asserting that throwing an exception is an unreachable path.

But I thought again, maybe it is also feasible to throw `IllegalArgumentException`, because Chunk may also be constructed by users themselves.

garydgregory replied:
Which leads to consider if we should present the smallest API surface for this new code, IOW the minimal elements that should be public or protected. Once public or protected, it's set in stone within a major release line.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImagingParameters.java:
```diff
@@ -0,0 +1,22 @@
+
+import org.apache.commons.imaging.common.XmpImagingParameters;
+
+public class WebPImagingParameters extends XmpImagingParameters {
```
Even small public methods and classes need a javadoc. It normally must include

- brief comment
- @since tag
- iff it's useful, tell users whether it's thread-safe or not
- if it contains complicated code, you can have a longer comment, include code examples, links to spec, etc.

kinow replied:
e.g. https://github.com/apache/commons-imaging/blob/cb0b52e5b82321cac1d6b4c97db7d58375322733/src/main/java/org/apache/commons/imaging/formats/png/PngImagingParameters.java#L23-L26

kinow reviewed (commented):
JIRA created: https://issues.apache.org/jira/browse/IMAGING-339

Haven't tested nor read the spec yet, but first wanted to have a slow read of the code to leave any comments. Once these are addressed I will test and try to verify correctness. But everything looks great! Should be included in the next alpha or beta release.

Thanks!
-Bruno

Glavo commented:
@kinow I went back to this work and completed the missing documents.

Can you take a look at my question here at  https://github.com/apache/commons-imaging/pull/254/files/08d6a0dff607eefa633c89d231a01623e3ad90c6#r1049312581? 
I feel that the usage of `ImageReadException` and `IOException` is often confused, so I don't know which one to choose.

Glavo commented:
I updated this PR, added the `@since` tag and more detailed javadoc.

kinow commented:
Sorry I am overseas visiting family until the end of September. Not sure if I will be able to review it until then.

spencerhakim commented:
Considering the libwebp CVE the industry is currently dealing with, would love to see this PR's conflicts resolved and this merged in sooner than later so I can move away from using a library that relies on JNI. 🙏

kinow commented:
> Considering the libwebp CVE the industry is currently dealing with, would love to see this PR's conflicts resolved and this merged in sooner than later so I can move away from using a library that relies on JNI. 🙏

@Glavo can you rebase and fix conflicts, please? I returned from my overseas trip last Friday. Just going through the backlog ($work/emails/etc.), but I should be able to review this pretty soon. Or if you are busy (and you trust us) just say so and we can try to fix the conflicts and review it. Thanks!

Glavo commented:
I've rebased and resolved conflicts.

I noticed that the project had changed a lot in my time away, so I simply updated this PR so that it would compile and pass tests. If there is anything else to update, it will take me some time to get back to this work.

kinow commented:
Thank you @Glavo ! Hopefully we can solve simple issues and we won't need any major refactoring for the review :crossed_fingers: Coming back to this one soon... (allowed GH Actions to run)

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageMetadata.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.List;
+
+/**
+ * @since 1.0-alpha4
```
Maybe the description here could be improved.

kinow replied:
Done.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
```diff
@@ -0,0 +1,326 @@
+import static org.apache.commons.imaging.common.BinaryFunctions.skipBytes;
+
+/**
+ * @since 1.0-alpha4
```
Maybe the description here could be improved, perhaps including a link to where this implementation came from (spec, docs, another library, etc.)

kinow replied:
Done.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
```diff
@@ -0,0 +1,326 @@
+        }
+
+        if (is.read(buffer) < 4 || !WebPConstants.WEBP_SIGNATURE.equals(buffer)) {
+            throw new ImagingException("Not a Valid WebP File");
```
Been too long since I worked on the code, but I **think** there is an `ImageReadException`. Have to confirm if other parsers raise that or this error, then try to keep it consistent.

kinow replied:
No, maybe there was, but it's doing exact same thing as other parsers. Resolving this comment.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunk.java:
```diff
@@ -0,0 +1,83 @@
+ * A WebP image is composed of several chunks. This is the base class for the chunks,
+ * used by the parser.
+ *
+ * @see <a href="https://developers.google.com/speed/webp/docs/riff_container">WebP Container Specification</a>
```
Maybe this should be documented in the parser as well.

Glavo replied:
None of the parsers for the other formats have javadoc, so I'm confused as to where to put the documentation.

kinow replied:
If the others are not doing that, then no worries @Glabo :) I spoke without looking at the other parsers (more a note as I did a first pass without the IDE). Thanks for replying it!

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunkALPH.java:
```diff
@@ -0,0 +1,41 @@
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * |Rsv| P | F | C |     Alpha Bitstream...                        |
+ * +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+ * }</pre>
```
:ok_man: Good docs! Thanks!

kinow commented on src/test/java/org/apache/commons/imaging/formats/webp/WebPBaseTest.java:
```diff
@@ -0,0 +1,39 @@
+import java.io.IOException;
+import java.util.List;
+
+public abstract class WebPBaseTest extends AbstractImagingTest {
```
Note to self: did we use (or started using) the since tag in tests too? Or was it in another component?

kinow replied:
No we did not. Resolving this comment.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
I went through the other files and they all look OK. Added some minor comments. This is the only class that seems it will take a bit longer to review (which is normal, the parsers are normally the most used classes, with more functions).

kinow reviewed (commented):
Had a quick look at the code, looks in excellent state. Left some comments, but I think I can fix most, if not all, during the review/merge.

Remains to review the parser code, and go through the specification & docs confirming the code is working as expected. Maybe test with exiftool (or similar tool), and check the tests coverage.

Thanks @Glavo !

garydgregory commented:
We don't need to use since tags IMO since we don't have a 1.0 yet.

kinow commented:
> We don't need to use since tags IMO since we don't have a 1.0 yet.

I think there are other since tags (can't recall if only for 1.0-alphaX or for 0.98 (sanselan) as well)

kinow reviewed (commented):
Rebasing onto `master` locally today to start reviewing it; hoping I can finish it this weekend.

I'm squashing all 23 commits into a single one, @Glavo. Here's the commits message I'm using (it will be attributed to your user; I used the text from your pull request, plus the prefix with the JIRA issue):

![image](https://github.com/apache/commons-imaging/assets/304786/374fffff-a6e3-433b-8912-465da143d700)

If I manage to review and it is ready to merge, I will proceed to push-force the squashed commits with this message :+1:

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunkVP8X.java:
```diff
@@ -0,0 +1,118 @@
+        return hasICC;
+    }
+
+    public boolean isHasAlpha() {
```
@Glavo would it OK to call this method "hasAlpha()" instead of "isHasAlpha" ? Ditto for the other methods? (I can batch rename them as I'm modifying the code, writing Javadocs, etc)

kinow replied:
Done! Maintaining the variable as `hasAlpha`.

kinow commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunkVP8X.java:
```diff
@@ -0,0 +1,118 @@
+        this.canvasWidth = (bytes[4] & 0xFF) + ((bytes[5] & 0xFF) << 8) + ((bytes[6] & 0xFF) << 16) + 1;
+        this.canvasHeight = (bytes[7] & 0xFF) + ((bytes[8] & 0xFF) << 8) + ((bytes[9] & 0xFF) << 16) + 1;
+
+        if (canvasWidth < 0 || canvasHeight < 0 || canvasWidth * canvasHeight < 0) {
```
I'll remove the `canvasWidth < 0 || canvasHeight < 0` from the if statement. IntelliJ had a code inspection warning here, and after some napkin bit math, I think `canvasWidth` and `canvasHeight` can not be negative. They can, however, be quite large, causing the multiplication to give a negative number.

Will see whether we should leave the `canvasWidth * canvasHeight < 0` and/or if it can be replaced by one of those JVM safe methods like `multiplyExact`.

kinow replied:
Not needed, resolving it.

kinow reviewed (commented):
Alright, done with the first code review, found no major issues.

I had a look at the public methods, and they look OK (returning the attributes of classes in getters/setters) just missing Javadocs that I will add later.

Modified one part that IntelliJ complained and it was safe to remove an `if` check (although we can add it back if it's more readable to others).

Next I will wait for @Glavo reply on the names of getters for booleans in some chunks, rename them if Glavo's OK with that, add javadoc to all public methods, ditto for new tests. Then the final review will be looking at the spec/docs linked in some classes, and at a google page that I remember reading some time ago with info about webp (or it could be a facebook post, I should have it in my bookmarks).

If there is nothing that stands out too different from the docs linked, this should be ready to be merged! :neckbeard: 

Thanks heaps for the great contribution @Glavo ! Sorry for the delay in reviewing it.

Glavo commented:
I probably just named them arbitrarily, so feel free to rename them if necessary.

kinow commented:
> I probably just named them arbitrarily, so feel free to rename them if necessary.

Thanks @Glavo !

Glavo commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunkVp8x.java:
```diff
@@ -42,76 +44,105 @@
-    private final boolean hasXmp;
-    private final boolean isAnimation;
+public final class WebPChunkVp8x extends WebPChunk {
+    private final boolean icc;
```
Should these field names be renamed? Referring to `PamFileInfo`, perhaps it would be better for us to keep the 'has' prefix in the field name.

kinow replied:
No strong opinion here, let me revert it.

Glavo commented on src/main/java/org/apache/commons/imaging/formats/webp/WebPImageParser.java:
```diff
@@ -277,13 +279,14 @@ public void close() throws IOException {
                 }
                 boolean padding = (payloadSize % 2) != 0;
-                int chunkSize = payloadSize + 8 + (padding ? 1 : 0);
+                int extraPadding = Math.addExact(8, (padding ? 1 : 0));
```
The maximum value of `extraPadding` is 9, so there is no risk of overflow.

Furthermore, the name `extraPadding ` is somewhat misleading, since `padding` only refers to the padding at the end of the chunk.

kinow replied:
>The maximum value of extraPadding is 9, so there is no risk of overflow.

`9` as per specification? Or is there a constant somewhere?

Asking as we had security issues that had to be fixed (with some urgency, which is no fun) where users crafted images with custom metadata, many times invalid image formats.

So we just need to confirm if `9` is the maximum value, or if the user can somehow manipulate it.

>Furthermore, the name extraPadding is somewhat misleading, since padding only refers to the padding at the end of the chunk.

Sorry, I just changed that, pushing a new version now... it uses a function to avoid having to delcare variables.

Glavo commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunk.java:
```diff
@@ -61,22 +74,41 @@ public String getTypeDescription() {
-        return 4 + 4 + size + padding;
+        int chunkFourCCandSize = 4 + 4;
+        int paddedSize = Math.addExact(size, padding);
+        return Math.addExact(chunkFourCCandSize, paddedSize);
```
I think we should check chunk size in constructor instead of here, this method should not throw exception.

kinow replied:
Good idea. Let me try that instead.

kinow replied:
Have a look if that's what you had in mind :+1: thanks!

Glavo commented on src/main/java/org/apache/commons/imaging/internal/SafeOperations.java:
```diff
@@ -0,0 +1,49 @@
+        if (values == null || values.length < 2) {
+            throw new IllegalArgumentException("You must provide at least two elements to be added");
+        }
+        return Arrays
```
It doesn't seem to be that easy to optimize. Maybe we should add a few overloads?

kinow replied:
I didn't think too much about this class yet. Wrote the first version that came to my mind without worrying about optimizing it. Maybe we can check if that becomes an issue and fix it if needed?

Glavo replied:
> I didn't think too much about this class yet. Wrote the first version that came to my mind without worrying about optimizing it. Maybe we can check if that becomes an issue and fix it if needed?

Maybe. It's really not that noticeable compared to the I/O overhead.

Glavo commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunk.java:
```diff
@@ -29,30 +30,50 @@
     private final int type;
     private final int size;
     protected final byte[] bytes;
+    private int chunkSize;
```
The `final` modifier should be added to it.

kinow replied:
Good catch! Done.

kinow reviewed (commented):
Pushed a new commit (thanks for marking it as OK for maintainers to update your branch, @Glavo ). Here's the summary of changes:

* Javadocs added to every public method;
* Renamed `hasAlpha` and similar boolean attributes to ~`alpha`, then the method became `hasAlpha()`~ (reverted following @Glavo 's feedback - thanks!);
* VP8, VP8L, and VP8X seem to disagree on `height`, `imageHeight`, and `canvasHeight`. I thought they were all the same, but the developers docs use these terms, so I think the current values are the best choice;
* Replaced some arithmetic operations by the JVM *exact methods, to avoid overflows and possible issues (especially security issues, fuzzing issues, etc.) - see `SafeOperations` class with a helper static method, in the **internal** package;
* WebP chunk implementations were throwing `ImagingException`, but the parent WebPChunk constructor was throwing `IllegalArgumentException`, but since the value is read from the image, I changed that to throw `ImagingException` too (the constructor actually declared `@throws ImagingException` but never raised it);
* Renamed classes like `WebPChunkALPH` to `WebPChunkAlph` (in other formats we have `BlablaExif`, `BlablaIcc`, etc.);
* Checked all headers and comments were OK.

Also resolved all pending comments, but I think we have one or two new conversations started just now :slightly_smiling_face: we should be able to solve them all pretty easily.

The final step will be to review the tests. I reviewed warnings raised by IntelliJ, and the last one I have now is about public methods never used. This probably indicates these public methods are not being called in unit tests too. Can't tell if important or not until I run the tests with coverage. Will do that over this week (maybe tomorrow).

Gotta go now, thanks @Glavo !

-Bruno

Glavo commented on src/main/java/org/apache/commons/imaging/formats/webp/chunks/WebPChunkVp8x.java:
```diff
@@ -42,76 +45,105 @@
-        this.canvasWidth = (bytes[4] & 0xFF) + ((bytes[5] & 0xFF) << 8) + ((bytes[6] & 0xFF) << 16) + 1;
-        this.canvasHeight = (bytes[7] & 0xFF) + ((bytes[8] & 0xFF) << 8) + ((bytes[9] & 0xFF) << 16) + 1;
+        this.canvasWidth = SafeOperations.add((bytes[4] & 0xFF), ((bytes[5] & 0xFF) << 8), ((bytes[6] & 0xFF) << 16), 1);
+        this.canvasHeight = SafeOperations.add((bytes[7] & 0xFF), ((bytes[8] & 0xFF) << 8), ((bytes[9] & 0xFF) << 16), 1);
```
There is no need to use `SafeOperations` here. Both `canvasWidth ` and `canvasHeight ` are 24-bit integers, so overflow is not possible.

I forget why I want to check that they are non-negative, but I guess using `SafeOperations` would mislead the reader that an overflow might occur here, so it might be better to revert it.

Glavo replied:
The other changes look good. I’m going to take a break too, and I’ll revisit the entire PR later to see what I haven’t forgotten.

kinow replied:
Thanks @Glavo ! Will update it over next day(s) :+1:

kinow replied:
You were right @glavo, reverted the change :+1:

kinow reviewed (commented):
Current webp package coverage, with JaCoCo,

![image](https://github.com/apache/commons-imaging/assets/304786/724acd8e-65bf-4b5c-a893-38773c341da2)


![image](https://github.com/apache/commons-imaging/assets/304786/9dd96c9f-54fd-4d2c-87ab-1d6f259f32cf)

It doesn't seem too hard to increase that coverage, so we can add a few more tests before merging it, not necessarily 100%, but we can definitely get it a little higher.

Other than that had a look at the docs and everything looks OK. So the pending now is:

- [x] add a few more simple tests
- [ ] write a quick java program to scan a directory and try to parse webp files (I have a bunch of life drawing reference that were saved as webp files, from multiple web sites and social media)
- [x] write in the `changes.xml` about the new format :tada:

garydgregory commented:
I hope we can get coverage above 90%.

kinow commented:
Well, this is as much as I could get today.

![image](https://github.com/apache/commons-imaging/assets/304786/d5686342-54d6-4bd2-a4b1-d0edd2f4d8d8)

![image](https://github.com/apache/commons-imaging/assets/304786/3467fc07-3ffa-466f-959a-a76fc81ad6a0)

kinow reviewed (approved):
Hi @Glavo ,

I think I am done with the review and review and the changes. A couple of comments:

- VP8X doesn't have a version number, and I couldn't find in the developer docs anything saying that it should have, but other chunks have it... I guess that chunk doesn't need it?
- ALPH chunk was never used. I added it to the enum of chunk types... or was it intentional to not add that one?

No blockers, I think. And here's the latest coverage report:

![image](https://github.com/apache/commons-imaging/assets/304786/8f12ccbb-6a94-4710-884b-c66a05f82f0c)

![image](https://github.com/apache/commons-imaging/assets/304786/df4d1784-1ecd-4ecd-8cc9-6855941db2a6)

So 89% (was 68%) line coverage, 80% (was 51%) branch coverage. Most of the lines not covered are error handling. And one chunk is completely not tested, the `ALPH`.

I think this is good to go. If @Glavo doesn't find anything wrong, then it should be ready to be merged. @garydgregory would you review it, since I added some code too, it'd be best if another person could have a look at it - if you have time.

Will add another commit with the change log.

Bruno

Glavo commented:
I re-checked this PR and it looks good.

Glavo commented:
> VP8X doesn't have a version number, and I couldn't find in the developer docs anything saying that it should have, but other chunks have it... I guess that chunk doesn't need it?

I implemented it in full compliance with the developer documentation. I'm not sure why there's no version number, but I guess it's not needed.

> ALPH chunk was never used. I added it to the enum of chunk types... or was it intentional to not add that one?

I don't remember the reason for this, so it should just be an oversight on my part. Thanks for the fix.

kinow commented:
Thank you @Glavo ! Let's wait a while to see if Gary has time to have a quick look just to see if I missed anything important. Then it should be ready to be merged. Thanks!!!
